### PR TITLE
Diff: Prevent invalid deletion

### PR DESCRIPTION
**Pull request by Techievena**
Prevent out of range deletion in Diff.
Raise error in case of out of range deletion.

Fixes https://github.com/coala/coala/issues/3782

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [X] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [X] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

File: coalib/results/Diff.py [hunk 1]
```diff
@@ -352,9 +353,12 @@ def delete_line(self, line_nr):
         """
         Mark the given line nr as deleted. The first line is line number 1.
         """
-        linediff = self._get_change(line_nr)
-        linediff.delete = True
-        self._changes[line_nr] = linediff
+        if (line_nr <= self._length):
```
Comment (Udayan12167) [inline coalib/results/Diff.py]: Try catch? Don't need to store an extra variable.

File: coalib/results/Diff.py [hunk 2]
```diff
@@ -352,9 +353,12 @@ def delete_line(self, line_nr):
...
+            linediff = self._get_change(line_nr)
```
Comment (Udayan12167) [inline coalib/results/Diff.py]: get_change already throws a value error? Why is another one needed

File: coalib/results/Diff.py (hunk 1 above)
Comment (Techievena) [inline coalib/results/Diff.py]: Sorry forgot to mark it WIP. I am currently working on this issue. I have already made changes. 😃

Comment (Udayan12167) [inline coalib/results/Diff.py]: I feel this should go in the _get_change method the check for length

File: coalib/results/Diff.py [hunk 3]
```diff
@@ -352,6 +352,7 @@ def delete_line(self, line_nr):
         """
         Mark the given line nr as deleted. The first line is line number 1.
         """
+        line_nr = 0 if (len(self._file) < line_nr) else line_nr
```
Comment (RohanVB) [inline coalib/results/Diff.py]: Why is the if-else like this? no new lines for formatting?

Comment (Techievena) [inline coalib/results/Diff.py]: @RohanVB Only a single value had to be assigned using if-else so I used this format to keep it short and simple :smiley:

Comment (RohanVB) [inline coalib/results/Diff.py]: Reasons why I don't like this format: 
1) It goes against the normal flow of thought (readability, familiarity).
2) Consistency in the codebase; I don't believe we use it elsewhere, for single values either. (In this file at least)
3) It doesn't follow the order of other languages (not really a major concern).

File: coalib/results/Diff.py [hunk 4]
```diff
@@ -352,6 +352,7 @@ def delete_line(self, line_nr):
         """
         Mark the given line nr as deleted. The first line is line number 1.
```
Comment (aptrishu) [inline coalib/results/Diff.py]: Document how it handles the non-existent line.

Comment (aptrishu) [inline coalib/results/Diff.py]: Also should we display proper error message in `delete_lines` after catching the error from `_get_changes`.

File: coalib/results/Diff.py (hunk 3 above)
Comment (Techievena) [inline coalib/results/Diff.py]: @RohanVB Well  I agree with you this format is not in use in this file, but this format is actually in use in many places in the codebase of coala.

Comment (RohanVB) [inline coalib/results/Diff.py]: alright :D

File: coalib/results/Diff.py [hunk 5]
```diff
@@ -351,14 +351,17 @@ def __bool__(self):
...
+        line_nr = 0 if (len(self._file) < line_nr) else line_nr
         linediff = self._get_change(line_nr)
         linediff.delete = True
         self._changes[line_nr] = linediff
 
     def delete_lines(self, line_nr_start, line_nr_end):
         """
-        Delete lines in a specified range, inclusively.
+        Delete lines in a specified range, inclusively. The range must be
+        valid, i.e. lines must exist in diff, else an error message is raised.
```
Comment (satwikkansal) [inline coalib/results/Diff.py]: `else error is raised.`

File: coalib/results/Diff.py [hunk 6]
```diff
@@ -351,14 +351,17 @@ def __bool__(self):
     def delete_line(self, line_nr):
         """
         Mark the given line nr as deleted. The first line is line number 1.
+        Raises an error if line nr doesn't exist in the diff.
         """
+        line_nr = 0 if (len(self._file) < line_nr) else line_nr
```
Comment (satwikkansal) [inline coalib/results/Diff.py]: Any reason for setting it to 0 :sweat_smile:  IMO you should raise an error here only.

File: coalib/results/Diff.py (hunk 5 above)
Comment (satwikkansal) [inline coalib/results/Diff.py]: Also where is the code that justifies
>lines must exist in diff, else an error message is raised

File: tests/results/DiffTest.py [hunk 7]
```diff
@@ -37,13 +37,15 @@ def test_delete_line(self):
...
 
     def test_delete_lines(self):
-        self.uut.delete_lines(1, 10)
-        self.uut.delete_lines(10, 20)
+        self.uut.delete_lines(1, 2)
+        self.uut.delete_lines(2, 3)
         additions, deletions = self.uut.stats()
-        self.assertEqual(deletions, 20)
-        self.assertRaises(ValueError, self.uut.delete_lines, 0, 10)
+        self.assertEqual(deletions, 3)
```
Comment (satwikkansal) [inline tests/results/DiffTest.py]: Wondering why to change 10,20 to 2,3. Please keep the original values if it doesn't test any case that might have been missed with previous values (10 and 20)

File: coalib/results/Diff.py (hunk 6 above)
Comment (Techievena) [inline coalib/results/Diff.py]: line nr 0 or less  than that is considered as invalid line number. Well for a diff of 5 lines, line nr 10 is also invalid line nr.

File: coalib/results/Diff.py (hunk 5 above)
Comment (Techievena) [inline coalib/results/Diff.py]: `delete_lines` calls `delete_line` which calls `_get_change`. and error is raised in `_get_change`

File: tests/results/DiffTest.py (hunk 7 above)
Comment (Techievena) [inline tests/results/DiffTest.py]: 10,20 is out of range for diff. That's why also I have added tests to handle cases like that. https://github.com/coala/coala/pull/3817/files#diff-8ab4f61ecc762c2ad5312562c454b2aaR48

File: coalib/results/Diff.py [hunk 8]
```diff
@@ -351,14 +351,17 @@ def __bool__(self):
...
         linediff = self._get_change(line_nr)
         linediff.delete = True
         self._changes[line_nr] = linediff
 
     def delete_lines(self, line_nr_start, line_nr_end):
         """
-        Delete lines in a specified range, inclusively.
+        Delete lines in a specified range, inclusively. The range must be
+        valid, i.e. lines must exist in diff, else an error is raised.
```
Comment (aptrishu) [inline coalib/results/Diff.py]: A newline here.

File: coalib/results/Diff.py [hunk 9]
```diff
@@ -351,14 +351,17 @@ def __bool__(self):
     def delete_line(self, line_nr):
         """
         Mark the given line nr as deleted. The first line is line number 1.
+        Raises an error if line nr doesn't exist in the diff.
```
Comment (aptrishu) [inline coalib/results/Diff.py]: A newline here, before `Raises an error if..`

**Comment by Techievena** [comment]
@aptrishu Is it Ok to be merged now?

File: coalib/results/Diff.py [hunk 10]
```diff
@@ -351,14 +351,20 @@ def __bool__(self):
     def delete_line(self, line_nr):
         """
         Mark the given line nr as deleted. The first line is line number 1.
+
+        Raises an error if line nr doesn't exist in the diff.
```
Comment (adtac) [inline coalib/results/Diff.py]: Uh, it doesn't raise an error? It just deletes the first line for some reason:

    line_nr = 0 if (len(self._file) < line_nr) else line_nr

Comment (adtac) [inline coalib/results/Diff.py]: Also, I think you want to use the word "exception".

File: coalib/results/Diff.py [hunk 11]
```diff
@@ -351,14 +351,20 @@ def __bool__(self):
...
         """
+        line_nr = 0 if (len(self._file) < line_nr) else line_nr
```
Comment (adtac) [inline coalib/results/Diff.py]: Ah you're setting it to zero which will raise an `ValueError`.

Why not simply do a `raise ValueError` if it's out of bounds? Actually, do a `raise IndexError` since that makes more sense.

File: coalib/results/Diff.py [hunk 12]
```diff
@@ -351,14 +351,20 @@ def __bool__(self):
     def delete_line(self, line_nr):
         """
         Mark the given line nr as deleted. The first line is line number 1.
+
+        Raises an exception if line nr doesn't exist in the diff.
         """
+        line_nr = 0 if (len(self._file) < line_nr) else line_nr
```
Comment (meetmangukiya) [inline coalib/results/Diff.py]: please rather use `line_nr > len(self._file)` :)

File: coalib/results/Diff.py [hunk 13]
```diff
@@ -97,7 +97,7 @@ def _get_change(self, line_nr, min_line=1):
         if not isinstance(line_nr, int):
             raise TypeError('line_nr needs to be an integer.')
         if line_nr < min_line:
-            raise ValueError('The given line number is not allowed.')
+            raise IndexError('The given line number is not allowed.')
```
Comment (meetmangukiya) [inline coalib/results/Diff.py]: Unrelated Change ?

File: coalib/results/Diff.py (hunk 12 above)
Comment (Techievena) [inline coalib/results/Diff.py]: Any specific reason for this ? :smile:

File: coalib/results/Diff.py (hunk 13 above)
Comment (Techievena) [inline coalib/results/Diff.py]: @meetmangukiya This change was asked by @adtac

File: coalib/results/Diff.py (hunk 12 above)
Comment (adtac) [inline coalib/results/Diff.py]: No braces around the if condition

Comment (adtac) [inline coalib/results/Diff.py]: Sorry, no parens around the if condition (unrelated to what @meetmangukiya said though)

Comment (Techievena) [inline coalib/results/Diff.py]: Didn't quite get you @adtac Am I supposed to add parenthesis instead of braces

File: coalib/results/Diff.py (hunk 13 above)
Comment (adtac) [inline coalib/results/Diff.py]: This should ideally be in its own commit since we want atomic commits. But if there are no other reviews, we can ignore it for now. (Although next time, you should make commits atomic.)

File: coalib/results/Diff.py [hunk 14]
```diff
@@ -351,14 +351,20 @@ def __bool__(self):
     def delete_line(self, line_nr):
         """
         Mark the given line nr as deleted. The first line is line number 1.
+
+        Raises an exception if line nr doesn't exist in the diff.
         """
+        line_nr = 0 if len(self._file) < line_nr else line_nr
```
Comment (adtac) [inline coalib/results/Diff.py]: I think I previously mentioned this in a previous review: just raise an exception directly if the line number is out of bounds. Don't make it zero and expect an exception elsewhere - that's not good practise.

**Comment by Techievena** [comment]
@adtac Done!!! :smile:

File: coalib/results/Diff.py [hunk 15]
```diff
@@ -351,14 +351,22 @@ def __bool__(self):
     def delete_line(self, line_nr):
         """
         Mark the given line nr as deleted. The first line is line number 1.
+
+        Raises an exception if line nr doesn't exist in the diff.
```
Comment (adtac) [inline coalib/results/Diff.py]: Either say `line_nr` with double backticks or use "line number"

File: coalib/results/Diff.py [hunk 16]
```diff
@@ -351,14 +351,22 @@ def __bool__(self):
...
         """
-        linediff = self._get_change(line_nr)
-        linediff.delete = True
-        self._changes[line_nr] = linediff
+        if line_nr > len(self._file):
+            raise ValueError('The given line number is out of bounds')
+        else:
```
Comment (adtac) [inline coalib/results/Diff.py]: There's no need for an else here. If an exception is raised, control would automatically go back to the calling function. Leave a line break.

File: tests/results/DiffTest.py [hunk 17]
```diff
@@ -37,13 +37,15 @@ def test_delete_line(self):
         additions, deletions = self.uut.stats()
         self.assertEqual(deletions, 1)
         self.assertRaises(ValueError, self.uut.delete_line, 0)
+        self.assertRaises(ValueError, self.uut.delete_line, 10)
```
Comment (adtac) [inline tests/results/DiffTest.py]: Shouldn't this be IndexError?

File: tests/results/DiffTest.py [hunk 18]
```diff
@@ -37,13 +37,15 @@ def test_delete_line(self):
...
-        self.uut.delete_lines(1, 10)
-        self.uut.delete_lines(10, 20)
+        self.uut.delete_lines(1, 2)
+        self.uut.delete_lines(2, 3)
         additions, deletions = self.uut.stats()
-        self.assertEqual(deletions, 20)
-        self.assertRaises(ValueError, self.uut.delete_lines, 0, 10)
+        self.assertEqual(deletions, 3)
+        self.assertRaises(ValueError, self.uut.delete_lines, 0, 2)
+        self.assertRaises(ValueError, self.uut.delete_lines, 1, 6)
```
Comment (adtac) [inline tests/results/DiffTest.py]: Wait, all these should be IndexError, not ValueError. How are the tests even passing?

File: coalib/results/Diff.py [hunk 19]
```diff
@@ -351,14 +351,22 @@ def __bool__(self):
     def delete_line(self, line_nr):
         """
         Mark the given line nr as deleted. The first line is line number 1.
+
+        Raises an exception if line number doesn't exist in the diff.
         """
+        if line_nr > len(self._file):
+            raise IndexError('The given line number is out of bounds')
```
Comment (meetmangukiya) [inline coalib/results/Diff.py]: Use a period at the end? Just suggesting because the TypeError above has a period at the end :P

**Comment by meetmangukiya** [comment]
Everything else looks good :)

File: coalib/results/Diff.py (hunk 19 above)
Comment (Techievena) [inline coalib/results/Diff.py]: Sure :+1: :smile:

File: coalib/results/Diff.py [hunk 20]
```diff
@@ -351,14 +351,22 @@ def __bool__(self):
     def delete_line(self, line_nr):
         """
         Mark the given line nr as deleted. The first line is line number 1.
+
+        Raises an exception if line number doesn't exist in the diff.
         """
+        if line_nr > len(self._file):
+            raise IndexError('The given line number is out of bounds.')
```
Comment (meetmangukiya) [inline coalib/results/Diff.py]: Use the same error message as above

**Comment by meetmangukiya** [comment]
Rebase

**Comment by meetmangukiya** [comment]
ack 92d2c9879ef23dfec0016998376d2a1c52f2ba51 
ack 49a5255ce44fc11be5a2c495fed3e83518b290e7

**Comment by meetmangukiya** [comment]
ack fca665f 40e228e

**Comment by meetmangukiya** [comment]
@adtac can you approve your review ?

**Comment by meetmangukiya** [comment]
@adtac can you approve your review ?

**Comment by meetmangukiya** [comment]
@rultor merge

**Comment by meetmangukiya** [comment]
ack 40e228e

**Comment by meetmangukiya** [comment]
ack 40e228ef91dc58f56f1e234ccc0a0034b48d4b7d

**Comment by meetmangukiya** [comment]
ack 6e932d16bdc099178571a56cb1da8d61e900d172
ack c82bc1789080dbe3370fd21d664e0c3f96db601e

**Comment by meetmangukiya** [comment]
@rultor mere

**Comment by meetmangukiya** [comment]
@rultor merge!
